modules/porous_flow/src/utils/PorousFlowVanGenuchten.C

```diff
@@ -609,4 +609,268 @@ d2saturationHys(Real pc,
   }
   return d2s;
 }
+
+Real
+cubic(Real x, Real x0, Real y0, Real y0p, Real x1, Real y1, Real y1p)
+{
+  const Real d = x1 - x0;
+  const Real d2 = Utility::pow<2>(d);
+  const Real mean = 0.5 * (x1 + x0);
+  const Real sq3 = 0.5 * std::sqrt(3.0) * d;
+  const Real term1 = y0p * (x - x0) * Utility::pow<2>(x - x1) /
+                     d2; // term1(x0) = term1(x1) = term1'(x1) = 0, term1'(x0) = y0p
+  const Real term2 = y1p * (x - x1) * Utility::pow<2>(x - x0) /
+                     d2; // term2(x0) = term2(x1) = term2'(x0) = 0, term2'(x1) = y1p
+  const Real term3 = (x - mean - sq3) * (x - mean) * (x - mean + sq3);
+  // term3' = (x - mean) * (x - mean + sq3) + (x - mean - sq3) * (x - mean + sq3) + (x - mean - sq3)
+  // * (x - mean)
+  //        = 3 (x - mean)^2 - sq3^2
+  // note term3' = 0 when x = mean +/- sq3/sqrt(3) = 0.5 * (x1 + x0) +/- 0.5 * (x1 - x0) = {x1, x0}
+  const Real term3_x0 = (x0 - mean - sq3) * (x0 - mean) * (x0 - mean + sq3);
+  const Real term3_x1 = (x1 - mean - sq3) * (x1 - mean) * (x1 - mean + sq3);
+  return (y0 * (term3 - term3_x1) + y1 * (term3_x0 - term3)) / (term3_x0 - term3_x1) + term1 +
+         term2;
+}
+
+Real
+dcubic(Real x, Real x0, Real y0, Real y0p, Real x1, Real y1, Real y1p)
+{
+  const Real d = x1 - x0;
+  const Real d2 = Utility::pow<2>(d);
+  const Real mean = 0.5 * (x1 + x0);
+  const Real sq3 = 0.5 * std::sqrt(3.0) * d;
+  const Real term1_prime = y0p * (Utility::pow<2>(x - x1) + (x - x0) * 2 * (x - x1)) / d2;
+  const Real term2_prime = y1p * (Utility::pow<2>(x - x0) + (x - x1) * 2 * (x - x0)) / d2;
+  const Real term3_prime =
+      3.0 * Utility::pow<2>(mean) - 6 * mean * x - 0.75 * d2 + 3.0 * Utility::pow<2>(x);
+  const Real term3_x0 = (x0 - mean - sq3) * (x0 - mean) * (x0 - mean + sq3);
+  const Real term3_x1 = (x1 - mean - sq3) * (x1 - mean) * (x1 - mean + sq3);
+  return (y0 - y1) * term3_prime / (term3_x0 - term3_x1) + term1_prime + term2_prime;
+}
+
+Real
+relativePermeabilityHys(Real sl,
+                        Real slr,
+                        Real sgrdel,
+                        Real sgrmax,
+                        Real sldel,
+                        Real m,
+                        Real upper_liquid_param,
+                        Real y0,
+                        Real y0p,
+                        Real y1,
+                        Real y1p)
+{
+  if (sl < slr) // by the definition of slr, always return 0
```

Comment:
Could this be `sl<= sir`? I think if `sl==slr` that `kr=0` too?

Reply:
oops - auto corrected `slr` to `sir`!

Reply:
Yea `<=` is better, i agree